cranelift/filetests/src/zkasm_codegen.rs

```diff
@@ -288,68 +290,154 @@ pub fn compile_clif_function(func: &Function) -> Vec<String> {
     // TODO: I believe it can be done more beautiful way
     let mut funcname = func.name.to_string();
     funcname.remove(0);
-    funcname.push(':');
-    let mut res = vec![funcname];
+    let mut res = vec![format!("{}:", funcname)];
     res.append(&mut lines);
+    res.into_iter()
+        .map(|s| s.replace("label", &format!("label_{}", funcname)))
+        .collect()
+}
+
+/// Builds main for test program
+pub fn build_main(invoke_names: Vec<String>) -> Vec<String> {
+    let mut res = vec![
+        "main:".to_string(),
+        "  SP - 1 => SP".to_string(),
+        "  RR :MSTORE(SP)".to_string(),
+    ];
+    for name in invoke_names {
+        res.push("  zkPC + 2 => RR".to_string());
+        res.push(format!("  :JMP({})", name));
+    }
+    res.push("  $ => RR :MLOAD(SP)".to_string());
+    res.push("  SP + 1 => SP".to_string());
+    res.push("  :JMP(RR)".to_string());
     res
 }
 
-// TODO: this function should be much rewrited,
-// now it works for one very basic case:
-// Simple progam which don't contain globals or some other speciefic preamble\postamble
-// Program don't need helper functions (for example 2-exp.zkasm)
-// How to fix it? Use generate_preamble and provide correct inputs for it.
-/// Builds zkASM used in filetests.
-pub fn build_test_zkasm(functions: Vec<Vec<String>>, invocations: Vec<Vec<String>>) -> String {
+/// Generate invoke name in format <function_name>_<arg>_..._<arg>
+pub fn invoke_name(invoke: &Invocation) -> String {
+    let mut res = invoke.func.clone();
+    for arg in &invoke.args {
+        res.push_str(&format!("_{}", arg));
+    }
+    res
+}
+
+/// Assembles all parts of zkasm test program together
+pub fn build_test_zkasm(
+    functions: Vec<Vec<String>>,
+    invocations: Vec<Vec<String>>,
+    main: Vec<String>,
+) -> String {
     // TODO: use generate_preamble to get preamble
     let preamble = "\
 start:
+  0xffff => SP
   zkPC + 2 => RR
     :JMP(main)
     :JMP(finalizeExecution)";
-    let mut main = vec![
-        "main:".to_string(),
-        "  SP - 1 => SP".to_string(),
-        "  RR :MSTORE(SP)".to_string(),
-    ];
-    for invocation in invocations {
-        main.extend(invocation);
-    }
-    main.push("  SP - 1 => SP".to_string());
-    main.push("  :JMP(RR)".to_string());
     let mut postamble = generate_postamble();
     let mut program = vec![preamble.to_string()];
-    program.append(&mut main);
+    program.extend(main);
+    for inv in invocations {
+        program.extend(inv);
+    }
     for foo in functions {
         program.extend(foo);
     }
     program.append(&mut postamble);
     program.join("\n")
 }
 
+fn runcommand_to_wat(invoke: Invocation, _compare: Comparison, expected: Vec<DataValue>) -> String {
+    // TODO: support different amounts of outputs
```

Comment:
FWIW we don't need to do this for near use-cases, multi-value is an extension over the wasm-core-1 spec and nearcore runtimes don't enable it. I guess an exception to this might be if 0 outputs are expected...